### ISSUE TITLE
Promote get_loss_preds_diffs to an op and simplify the schemas that exist only because it is not one

### ISSUE THREAD

**Issue by speediedan** [speediedan/interpretune#282]
Promote `get_loss_preds_diffs` to a first-class analysis op, then simplify the schemas that exist only because it is not one.

`ops/bundled/attribution/attribution_ops.yaml` binds it as an `importable_params` callable (lines 16 and 84, `interpretune.analysis.optools.get_loss_preds_diffs`) and carries a standing TODO at line 86:

> remove appropriate output_schema input_schema columns if/when get_loss_preds_diffs is made an op

So the current shape is understood to be provisional by the file itself. Two consequences worth stating:

- **The schemas carry columns they would not need.** Because the loss/preds/diffs computation is smuggled in as an importable callable rather than declared as a required op, the input and output schemas of its dependents have to name columns that exist to feed it. Those are exactly the columns the TODO says to remove.
- **It is invisible to op-level machinery.** An `importable_params` binding does not participate in `required_ops`, schema compilation, precedence resolution, `op_info`, or stub generation the way an op does. Anything reasoning about the op graph cannot see it.

## Scope

- [ ] Declare `get_loss_preds_diffs` as an op (family placement: `attribution` is where its callers live, but `core` may be the better home if non-attribution ops want it)
- [ ] Convert its dependents to `required_ops` rather than an `importable_params` binding
- [ ] Remove the now-unnecessary `input_schema`/`output_schema` columns the TODO names, and delete the TODO
- [ ] Regenerate stubs (`scripts/generate_op_stubs.py`) and confirm a clean `git diff`
- [ ] Confirm the analysis parity gates are unmoved — this should be a pure restructuring with identical numerics

## Interaction worth checking first

**#161** (op-driven transitive dependency atomicity) is live and directly relevant: converting an `importable_params` binding into a `required_ops` edge is exactly the operation whose compiled-schema behavior #161 says is too coarse. Measured on `main`, `get_alive_latents` compiles to `{answer_indices, cache, input}` while `get_alive_latents_impl(module, analysis_batch, batch_idx)` cannot receive `input`. Promoting an op here may add a dependent with the same mismatch, so whoever takes this should decide whether it wants #161 first or can be done independently.

_Filed from the 2026-08-19 pre-#261 queue staleness audit, as the actionable residual of #46 (closed as unactionable: it asked to switch a "test op" to something that is not an op, and the `labels_to_ids` it named is an overridden module method at `rte_boolq.py:307`, not the op)._

**Comment by speediedan** [comment]
> **CORRECTED 2026-08-20: it is FOUR sites, not three.** This comment said three; the issue body says two. Both undercounts came from grepping line numbers by eye. Parsing the YAML and asking which ops carry `get_loss_preds_diffs` under `importable_params` gives four, with no judgement involved:
>
> | file | op |
> | --- | --- |
> | `attribution/attribution_ops.yaml` | `model_gradient` |
> | `attribution/attribution_ops.yaml` | `ablation_attribution` |
> | `core/core_ops.yaml` | `logit_diffs` |
> | `core/core_ops.yaml` | `logit_diffs_cache` |
>
> The one this comment missed is the **second declaration in `core_ops.yaml`** (`logit_diffs_cache`). Credit to the session that re-derived it rather than taking my number.
>
> **Why eyeballing failed here, since it failed twice:** `get_loss_preds_diffs` appears **18 times** across those files as imports, impl signature defaults, call sites and comments. Counting *declarations* in that noise is a judgement call; loading the YAML is not. For any "find all N" question on these files, parse them.
>
> **The tree also carries three TODO markers naming further scope**, which neither the body nor this comment mentioned:
>
> ```
> core/core_ops.yaml:100
> core/core_ops.yaml:137
> attribution/attribution_ops.yaml:86
>     # TODO: remove appropriate output_schema input_schema columns
>     #       if/when get_loss_preds_diffs is made an op
> ```
>
> So converting the bindings is not the whole job: whoever wrote those expected schema columns to come out too, and marked where. Note **three markers against four declarations**, so one op's schema consequence is unmarked and will need working out.
>
> Net: the scope is larger than the body states, but it is already enumerated in the tree rather than needing rediscovery.

---

~~**Correction to this issue's body (2026-08-20 premise check): `get_loss_preds_diffs` is bound at THREE sites, not two.**~~ **See above: four.**

The body says lines 16 and 84 of `attribution_ops.yaml`. That undercount stands as a finding; the corrected total is four, and the fourth is `logit_diffs_cache` in `core_ops.yaml`.

**On sequencing:** #161 has landed (`main` @ `e6df61b`, via #289), so the hazard that motivated ordering this issue behind it — the `importable_params`-to-`required_ops` conversion manufacturing a spurious requirement, exactly the over-declaring #161 called too coarse — should now be retired. That should be **verified against the tree rather than assumed**, since it is the premise the ordering rested on.

**Comment by speediedan** [comment]
## Diagnosis: not implementable as written. Two of the four dependents structurally cannot use an op.

Measured against `main` before writing anything, since the issue's own "check first" section asks for
exactly this.

### The #161 interaction is retired

#161 landed (`2bc26d8`). Inherited inputs now compile to `required=False`, so converting an
`importable_params` binding into a `required_ops` edge no longer manufactures a requirement the
dependent cannot satisfy. **That hazard is gone and is not what blocks this.**

### Scope correction: four declarations, three implementations

The body says lines 16 and 84. Parsing the bundled YAMLs for ops carrying `get_loss_preds_diffs` under
`importable_params` gives **four**, backed by **three** implementations:

| op | implementation | how it uses the function |
| --- | --- | --- |
| `logit_diffs` | `logit_diffs_impl` | batch-level |
| `logit_diffs_cache` | `logit_diffs_impl` (same impl) | batch-level |
| `model_gradient` | `model_gradient_impl` | **traced backward closure** + batch-level (2 calls) |
| `ablation_attribution` | `ablation_attribution_impl` | **per-latent, inside a nested loop** |

Worth parsing rather than grepping: the symbol appears **18 times** across those files as imports,
signature defaults, call sites and comments.

### Why two of them cannot become `required_ops`

**`model_gradient` — the call is inside a closure handed to the backend.**

```python
def backward_fn(raw_logits: torch.Tensor) -> torch.Tensor:
    sliced = raw_logits[torch.arange(raw_logits.size(0)), answer_indices]
    squeezed = torch.squeeze(sliced, dim=1)
    _, logit_diffs, _, _ = get_loss_preds_diffs(module, analysis_batch, squeezed, logit_diff_fn)
    return logit_diffs.sum()

raw_logits = model_backend.fwd_w_grads_and_latent_models(..., backward_fn=backward_fn)
```

The existing comment states the constraint: *"Applied to raw logits inside the backend. Must use only
standard PyTorch ops so NNsight can trace through it (all operations intercepted via
`__torch_function__`)."*

A dispatched op cannot go there. It runs capability validation and input-schema validation, binds a
resolution context, updates and returns an `AnalysisBatch` — none of that is traceable, and the
closure must return a **scalar** for backprop, not a batch.

**`ablation_attribution` — the call is per-latent inside a nested loop**, over
`analysis_batch.answer_logits.items()` and then `alive_latents[act_name]`, consuming the 4-tuple for
per-latent metrics. An op is a batch-level, schema-validated call that updates the batch; invoking one
per latent is neither its shape nor its cost profile.

**Also: the argument is not a column.** At every site the third argument is a *derived local* —
`analysis_batch.answer_logits` indexed by `answer_indices` and squeezed, or raw logits sliced inside
the backward pass. It is never a stored field, so an op would either have to absorb the indexing
(different at each site) or have the intermediate written to the batch first, which **adds** a column
rather than removing one.

### The TODO's goal is not reachable for those two either

The columns exist because the function reads `analysis_batch.label_ids` and `analysis_batch.orig_labels`.
Those reads happen wherever it is called. Since `model_gradient` and `ablation_attribution` must keep
calling it directly, their schemas still need those columns, so
*"remove appropriate output_schema input_schema columns"* cannot be completed for them.

### Recommendation

**Re-scope or close.** The issue's premise — that this is a binding-to-edge conversion — holds for one
implementation and not the other two. Options:

1. **Narrow to `logit_diffs_impl`** (covering `logit_diffs` and `logit_diffs_cache`): promote the op,
   convert those two to `required_ops`, drop their now-inherited columns. Leaves the function in place
   for the two sites that need it, and leaves their TODOs standing with a note saying why. Real but
   partial, and it creates two ways to do the same thing.
2. **Close as not-actionable in this form**, and file the narrower piece separately if wanted.
3. **Restructure the callers first** so the traced closure and the per-latent loop no longer need it —
   a much larger change touching the NNsight backward path, and not obviously desirable.

My recommendation is **(1) or (2) over (3)**, and I would want a maintainer to choose rather than
assume: (1) buys tidier schemas for two ops at the cost of two mechanisms coexisting, which is a
judgement about the codebase's direction rather than a technical question.

**Comment by speediedan** [comment]
## Deferred by the maintainer, leaning toward closure

Decision (2026-08-20): **defer, with a lean toward closing**, and re-evaluate somewhere after #261
merges. Moved down the IT MVP board accordingly — repositioned from board slot 4 to immediately after
#261, so it is re-considered in that context rather than popped off the top of the queue.

Not implemented, and deliberately so: the [diagnosis above](https://github.com/speediedan/interpretune/issues/282#issuecomment-5363372264)
found the issue is not implementable as written. Two of the four dependents structurally cannot use an
op — `model_gradient` calls the function inside a backend-traced `backward_fn` closure that must return
a scalar and use only traceable PyTorch ops, and `ablation_attribution` calls it per-latent inside a
nested loop. Neither is a shape a dispatched, schema-validated op can take.

The `#161` interaction this issue asked to check first **is** retired, so that is no longer a reason to
wait; the blocker is the call-site shapes, which #261 does not change either. Re-evaluating post-#261
is about **priority**, not about expecting the obstacle to clear.

When it is re-evaluated, the live options are in the diagnosis: narrow to `logit_diffs_impl` only
(tidier schemas for two ops, at the cost of two mechanisms coexisting), or close. Restructuring the
NNsight backward path to accommodate an op was judged not worth it.

**Comment by speediedan** [comment]
**The body undercounts the binding, and the correction currently lives only in the comment thread. Recording it here so the top of the issue is not the thing a reader trusts.**

The body says the `importable_params` binding is at "lines 16 and 84" of `attribution_ops.yaml` — two sites. Parsing both YAMLs on `main` (parsing, not grepping — eyeballing undercounted twice before), there are **four**, across two files, all mapping to `interpretune.analysis.optools.get_loss_preds_diffs`:

| File | Op |
| --- | --- |
| `core/core_ops.yaml` | `logit_diffs` |
| `core/core_ops.yaml` | `logit_diffs_cache` |
| `attribution/attribution_ops.yaml` | `model_gradient` |
| `attribution/attribution_ops.yaml` | `ablation_attribution` |

All three TODO markers also survive (`core_ops.yaml:100`, `:137`, `attribution_ops.yaml:86`), and `get_loss_preds_diffs` is still not declared as an op anywhere under `ops/bundled/`.

**Two things that have changed since filing:**

1. The **#161 interaction is genuinely retired** — #161 is closed, `2bc26d8` is on `main`, and the inherited-input merge at `schema_compiler.py:447-457` sets `inherited_input["required"] = False`. So that hazard no longer blocks anything; it is the call-site *shapes* that do, exactly as the diagnosis said.
2. The stated re-evaluation trigger has **not** fired: #261 is still open.

**This is a pending decision, not undelivered work**, which is worth stating plainly since it changes who can close it:

- **Narrow it** — promote `logit_diffs_impl` only, convert `logit_diffs` + `logit_diffs_cache` to `required_ops`, drop their now-inherited columns, and leave the two structurally-incompatible sites (the traced `backward_fn` closure at `attribution_ops.py:65-68`, and the per-latent call inside the `alive_latents` loop at `:220-221`) with their TODOs annotated as to why. Regenerate stubs; confirm parity gates unmoved.
- **Close as not-actionable in this form** and file the narrow piece separately.

Option 3 (restructuring the NNsight backward path) was already judged not worth it. Nothing in the current queue touches `analysis/ops/`, so neither option has been taken.
